net/net.go

```diff
@@ -154,3 +156,49 @@ func ParseIPCIDRs(cidrs []string) (*netipx.IPSet, error) {
 
 	return ips, nil
 }
+
+// SchemeHost parses URI string (without #fragment part) and returns schema used in this URI as first return value and
+// host[:port] part as second return value. Port is never omitted for HTTP(S): if no port is specified in URI, default port for given
+// schema is used. If URI is invalid, error is returned.
+// SchemeHost does not work with IPv6 addresses.
+func SchemeHost(input string) (string, string, error) {
+	u, err := url.ParseRequestURI(input)
+	if err != nil {
+		return "", "", err
+	}
+	if u.Scheme == "" {
+		return "", "", fmt.Errorf(`parse %q: missing scheme`, input)
+	}
+	if u.Host == "" {
+		return "", "", fmt.Errorf(`parse %q: missing host`, input)
+	}
+
+	// endpoint address cannot contain path, the rest is not case sensitive
+	s, h := strings.ToLower(u.Scheme), strings.ToLower(u.Host)
+
+	hh, p, err := net.SplitHostPort(h)
+	if err != nil {
+		// what is the actual right way of doing this, considering IPv6 addresses, too?
```

Comment:
What does the question mean?
We should be ipv6 safe

Reply:
This is a copy of existing implementation.
I think what is meant here is that it does not handle IPv6 without port case.

From https://www.ietf.org/rfc/rfc3986.txt
```
Appendix A.  Collected ABNF for URI

   URI           = scheme ":" hier-part [ "?" query ] [ "#" fragment ]

   hier-part     = "//" authority path-abempty
                 / path-absolute
                 / path-rootless
                 / path-empty

   URI-reference = URI / relative-ref

   absolute-URI  = scheme ":" hier-part [ "?" query ]

   relative-ref  = relative-part [ "?" query ] [ "#" fragment ]

   relative-part = "//" authority path-abempty
                 / path-absolute
                 / path-noscheme
                 / path-empty

   scheme        = ALPHA *( ALPHA / DIGIT / "+" / "-" / "." )

   authority     = [ userinfo "@" ] host [ ":" port ]
   userinfo      = *( unreserved / pct-encoded / sub-delims / ":" )
   host          = IP-literal / IPv4address / reg-name
   port          = *DIGIT

   IP-literal    = "[" ( IPv6address / IPvFuture  ) "]"

   IPvFuture     = "v" 1*HEXDIG "." 1*( unreserved / sub-delims / ":" )

   IPv6address   =                            6( h16 ":" ) ls32
                 /                       "::" 5( h16 ":" ) ls32
                 / [               h16 ] "::" 4( h16 ":" ) ls32
                 / [ *1( h16 ":" ) h16 ] "::" 3( h16 ":" ) ls32
                 / [ *2( h16 ":" ) h16 ] "::" 2( h16 ":" ) ls32
                 / [ *3( h16 ":" ) h16 ] "::"    h16 ":"   ls32
                 / [ *4( h16 ":" ) h16 ] "::"              ls32
                 / [ *5( h16 ":" ) h16 ] "::"              h16
                 / [ *6( h16 ":" ) h16 ] "::"

   h16           = 1*4HEXDIG
   ls32          = ( h16 ":" h16 ) / IPv4address
   IPv4address   = dec-octet "." dec-octet "." dec-octet "." dec-octet
```
It follows that `https://[::0]/foo/bar` IPv6 URI is valid.
The host `[::0]` would be and `net.SplitHostPort` will fail due to missing port.

I think we should trim `[]` in case of `missing port in address` error, see https://go.dev/play/p/iiRpN82LMnA

Reply:
Hmm Trim feels like a poor men's solution. I would guess that there is a netip or net package function that covers the problem.

Reply:
Trim is super fast, from "missing port in address" we know address is ok but the port is missing.

Reply:
Maybe https://pkg.go.dev/net/netip#ParseAddr ?

Reply:
> address is ok

Of course we don't validate hostname/address so this will pass `https://---invalid---:80/foo/bar` even though hostname is invalid. I don't think we should validate hostname/address here, it should fail later during request.

Reply:
Yes, we can use `netip.ParseAddr` to handle IPv6's properly. However, how should `SchemeHost` behave in case of DNS names? I was thinking it should consider DNS names valid and return host the same way it does with IPv4's and IPv6's.

Reply:
Lets add test cases for ipv4 and ipv6 with and without port

Reply:
Added them.